File: 16-3sum-closest/3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums.sort((a, b) => a - b)
    let closest = Infinity

    for(let i = 0; i < nums.length; i++){
        if(i > 0 && nums[i] === nums[i - 1]) continue

        let l = i + 1, r = nums.length - 1
        while(l < r){
            total = nums[i] + nums[l] + nums[r]
            current_dist = Math.abs(total - target)
            closest_dist = Math.abs(target - closest)

            if(current_dist < closest_dist){
                closest = total
            }

            if (total > target) r--
            else if (total < target) l ++
            else return total 
        }
    }
    return closest
};

File: 16-3sum-closest/3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        dist = float('inf')
        nums.sort()

        for i, num in enumerate(nums):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            l, r = i + 1, len(nums) - 1
            while l < r:
                total = num + nums[l] + nums[r]
                curr_dist = abs(target - total)
                closest_dist = abs(target - dist)

                if curr_dist < closest_dist:
                    dist = total 

                if total > target:
                    r -= 1
                elif total < target:
                    l += 1
                else:
                    return total
            
        return dist

        
        